crates/core/transaction/src/plan/build.rs

```diff
@@ -488,13 +235,149 @@ impl UnauthTransaction {
         }
 
         // Compute the binding signature and assemble the transaction.
-        let binding_signing_key = rdsa::SigningKey::from(self.synthetic_blinding_factor);
-        let auth_hash = self.inner.transaction_body.auth_hash();
+        let binding_signing_key = rdsa::SigningKey::from(synthetic_blinding_factor);
+        let auth_hash = transaction.transaction_body.auth_hash();
+
         let binding_sig = binding_signing_key.sign(rng, auth_hash.as_bytes());
         tracing::debug!(bvk = ?rdsa::VerificationKey::from(&binding_signing_key), ?auth_hash);
 
-        self.inner.binding_sig = binding_sig;
+        transaction.binding_sig = binding_sig;
+
+        Ok(transaction)
+    }
+
+    /// Build the serial transaction this plan describes.
+    /// Arguments
+    ///     - `full_viewing_key`, the [`FullViewingKey`] for the source funds;
+    ///     - `witness_data`, the [`WitnessData`] used for proving;
+    ///     -  `auth_data`, the [`AuthorizationData`] required to sign transaction.
+    pub fn build(
+        self,
+        full_viewing_key: &FullViewingKey,
+        witness_data: WitnessData,
+        auth_data: &AuthorizationData,
+    ) -> Result<Transaction> {
+        // Add the memo.
+        let mut _memo: Option<MemoCiphertext> = None;
+        let mut memo_key: Option<PayloadKey> = None;
+        if self.memo_plan.is_some() {
+            let memo_plan = self
+                .memo_plan
+                .clone()
+                .ok_or_else(|| anyhow!("missing memo_plan in TransactionPlan"))?;
+            _memo = memo_plan.memo().ok();
+            memo_key = Some(memo_plan.key);
+        }
+
+        let mut actions: Vec<Action> = Vec::new();
+
+        // We build the actions sorted by type, with all spends first, then all
+        // outputs, etc.  This order has to align with the ordering in
+        // TransactionPlan::effect_hash, which computes the auth hash of the
+        // transaction we'll build here without actually building it.
+
+        // 1. Build each action.
+        for spend_plan in self.spend_plans() {
+            let spend = ActionPlan::Spend(spend_plan.to_owned());
+            let action = spend
+                .build_unauth(full_viewing_key, &witness_data, memo_key.clone())
+                .expect("Build spend action failed!");
+            actions.push(action);
+        }
+        for output_plan in self.output_plans() {
+            let output = ActionPlan::Output(output_plan.to_owned());
+            let action = output
+                .build_unauth(full_viewing_key, &witness_data, memo_key.clone())
+                .expect("Build output action failed!");
+            actions.push(action);
+        }
+
+        // TODO: Handle other actions (swaps, swap claims, etc.).
```

Comment:
Rather than having to handle each action type specially, this should just run over the action plans in one go:
```
for action_plan in &self.action_plans {
    actions.push(action_plan.build_unauth(full_viewing_key, &witness_data, &memo_key));
}
```

Reply:
much cleaner!